### ISSUE TITLE
Command palette gets re-rendered even if hidden

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#13677]
## Description

Command palette gets re-rendered even if hidden.

## Reproduce

1. Open command palette and two tabs
2. Close the command palette
3. Go to inspector right click on command palette element, select "break on" > "subtree modifications"
4. Switch tabs
5. See that command palette fragments get re-rendered via lumino virtual DOM

![Screenshot from 2022-12-29 13-23-55](https://user-images.githubusercontent.com/5832902/209961278-164071f9-3032-4091-af3e-84102658cbcb.png)

It is not clear to me why this update appears. Of note the migration to `requestAnimationFrame` in lumino seems to make the debugging/profiling a bit more confusing as I don't immediately see if this was a genuine animation frame request or just lumino messaging.

## Expected behavior

Updating command palette on tab switch is expected because different commands will become available in different contexts (e.g. text editor/notebook), but re-rendering it while it is hidden is not expected.

In a bigger picture we could ask: should command palette get detached rather than hidden. I think hiding is ok because there is ever only one command palette, but requires a better hygiene on when a re-render is triggered.

## Context

- JupyterLab version: 4.0.0a31

**Comment by krassowski** [comment]
This also happens in 3.x branch.

**Comment by krassowski** [comment]
This could be solved by adding `this.isHidden` check in lumino [here](https://github.com/jupyterlab/lumino/blob/fa048e1680cce138af24ee1aa964f2f02b5e61f6/packages/widgets/src/commandpalette.ts#L289) or by detaching the command palette when it gets hidden. Again, detaching and re-attaching elements has a cost and this is a modal element which we want to show up immediately so I think it is ok to stay with current "always attached" behaviour.

**Comment by alec-kr** [comment]
Hey there. Is it possible to resolve this issue by nesting [L338](https://github.com/jupyterlab/lumino/blob/fa048e1680cce138af24ee1aa964f2f02b5e61f6/packages/widgets/src/commandpalette.ts#L338) in an "if" statement?

Something like the following:
```
if (!this.isHidden) {
    VirtualDOM.render(content, contentNode);
}
```
Or, should the entire body of the function be placed in this check, to prevent any unnecessary updating to hidden elements?

I would like to work on this issue, and some guidance will be appreciated 😄

**Comment by krassowski** [comment]
Yes, this could be as easy as that. I would try short-circuiting earlier up in the function to avoid doing any work if the palette is hidden.

**Comment by alec-kr** [comment]
Okay great. Can I perhaps create a PR including this change?

**Comment by alec-kr** [comment]
@krassowski Can you kindly re-check if the bug still occurs now that your suggested change has been made?

**Comment by krassowski** [comment]
Thank you @alec-kr and thank you for your patience. I have just checked with the latest JupyterLab development version (beta 0) and this still occurs because `this.isHidden` is always false for command palette when used in `ModalCommandPalette` which uses composition pattern:

https://github.com/jupyterlab/jupyterlab/blob/ee06f972d274fc0d7b3422af937b6e7a7d84894a/packages/apputils/src/commandpalette.ts#L15-L23

https://github.com/jupyterlab/jupyterlab/blob/ee06f972d274fc0d7b3422af937b6e7a7d84894a/packages/apputils/src/commandpalette.ts#L38-L48

This is because hiding the parent (`ModalCommandPalette`) does not get propagated down to the child (`CommandPalette`) in lumino.

As in https://github.com/jupyterlab/jupyterlab/pull/14172 we would need to check if any of the parents of command palette is hidden:

https://github.com/jupyterlab/jupyterlab/blob/ee06f972d274fc0d7b3422af937b6e7a7d84894a/packages/running/src/index.tsx#L307-L335

If `isAnyHidden()` turns out to be a frequently needed, maybe we should have this as a method of `Widget` in lumino?

**Comment by alec-kr** [comment]
Yes. I think it should be a method of ```Widget``` if it will be needed often.
